### PR TITLE
Issue 3402 : Updated section Running Pravega in Standalone Mode with SSL/TLS Enabled in  run-local.md

### LINKED ISSUE #3402: Document Update: Pravega Security documentation
**Problem description**
Updating documentation about setting up TLS for standalone mode

**Problem location**
documentation/pravega-security-configurations.md
documentation/deployment/run-local.md

**Suggestions for an improvement**
https://github.com/pravega/pravega/pull/3399

### PR DESCRIPTION
**Pull request by Aparnarr**
**Change log description**  
1. Major content edits in the section Standalone

**Purpose of the change**  
 Fixes #3402

### CODE REVIEW

**Comment by Aparnarr** [comment]
@ravisharda requested for review

File: documentation/src/docs/deployment/run-local.md [hunk 1]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
+  ```
+These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
 
 ## Running standalone with encryption and authentication enabled
 
 The configurations, `singlenode.enableTls` and `singlenode.enableauth` can be used to enable encryption and authentication respectively.(By default both these settings are disabled).
 
 In case `enableTls` is set to true, the default certificates provided in the `conf` directory are used for setting up TLS. These can be overridden by specifying in the properties file.
 
+1. Ensure that the server's certificate is trusted. If you run `/gradlew startStandalone` without it, you'll encounter the following error:
```
Comment (andreipaduroiu) [inline documentation/src/docs/deployment/run-local.md]: It's `./gradlew startStandalone` (you need the `.` at the beginning)

**Comment by Aparnarr** [comment]
@fpj assign @ravisharda as reviewer

File: documentation/src/docs/deployment/run-local.md (hunk 1 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @andreipaduroiu Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 2]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
+  singlenode.trustStoreJKS=../config/standalone.truststore.jks
+
+  ```
+These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
 
 ## Running standalone with encryption and authentication enabled
 
 The configurations, `singlenode.enableTls` and `singlenode.enableauth` can be used to enable encryption and authentication respectively.(By default both these settings are disabled).
 
 In case `enableTls` is set to true, the default certificates provided in the `conf` directory are used for setting up TLS. These can be overridden by specifying in the properties file.
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Say something along the lines of the following instead: 

In case you intend to enable SSL/TLS, set the following configurations: 
(paste the block of TLS related config here.)

These configurations use the provided certificates, keys and keystores, by default. You can supply your own instead if you want.

**Comment by Aparnarr** [comment]
@ravisharda will edit

File: documentation/src/docs/deployment/run-local.md [hunk 3]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
+  singlenode.keyFile=../config/key.pem
+  singlenode.certFile=../config/cert.pem
+  singlenode.keyStoreJKS=../config/standalone.keystore.jks
+  singlenode.keyStoreJKSPasswordFile=../config/standalone.keystore.jks.passwd
+  singlenode.trustStoreJKS=../config/standalone.truststore.jks
+
+  ```
+These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
 
 ## Running standalone with encryption and authentication enabled
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Should the title be something like "Running Pravega in standalone mode with SSL/TLS enabled?"
* The term `standalone` is too general a term. What we are referring to here is `Pavega in standalone mode` (borrowing from https://github.com/pravega/pravega/blob/master/documentation/src/docs/getting-started.md). 
* Everything we are talking about in this section is only related to SSL/TLS. 

Maybe we should all add another section at the same level titled: "Running Pravega in standalone mode with Client Authentication and Authorization enabled", adding auth related configurations.

File: documentation/src/docs/deployment/run-local.md [hunk 4]
```diff
@@ -9,15 +9,15 @@ You may obtain a copy of the License at
 -->
 # Running Pravega
 
-Running in local machine allows us to get started using Pravega very quickly. Most of the options uses the standalone mode which is suitable for most of the development and testing.
+Running in local machine allows us to get started using Pravega very quickly. Most of the options uses the standalone mode which is suitable for most of the development and testing purposes.
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 5]
```diff
@@ -9,15 +9,15 @@ You may obtain a copy of the License at
...
 
 The prerequisites for running in local machine is described below.
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 6]
```diff
@@ -9,15 +9,15 @@ You may obtain a copy of the License at
...
 
 ## Standalone Mode
 
 ### From Source
 
-First we need to have the Pravega source code checked out to download the dependencies:
+The Pravega source code needs to be checked out to download the dependencies:
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 7]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
+We need to set the IP address of our local machine as the value of `HOST_IP` in the following command.
 ```
 HOST_IP=1.2.3.4 docker-compose up
 ```
 Clients can then connect to the controller at `${HOST_IP}:9090`.
 
 ## Configuring standalone
 
-The configuration properties for standalone can be specified in `config/standalone-config.properties`.
+Configure standalone server to communicate using SSL/TLS. To do so, edit the TLS-related properties in `standalone-config.properties` as shown below:
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda  Please check the section Title

File: documentation/src/docs/deployment/run-local.md [hunk 8]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
-These properties include ports for zookeeper, segment store and controller. They also contain other configurations related to security.
+  ```java
+  singlenode.enableTls=true
+  singlenode.keyFile=../config/key.pem
+  singlenode.certFile=../config/cert.pem
+  singlenode.keyStoreJKS=../config/standalone.keystore.jks
+  singlenode.keyStoreJKSPasswordFile=../config/standalone.keystore.jks.passwd
+  singlenode.trustStoreJKS=../config/standalone.truststore.jks
+
+  ```
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda chnaged the section title

File: documentation/src/docs/deployment/run-local.md [hunk 9]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
+These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
 
 ## Running standalone with encryption and authentication enabled
 
 The configurations, `singlenode.enableTls` and `singlenode.enableauth` can be used to enable encryption and authentication respectively.(By default both these settings are disabled).
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 2 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 10]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
 
 In case `enableTls` is set to true, the default certificates provided in the `conf` directory are used for setting up TLS. These can be overridden by specifying in the properties file.
 
+1. Ensure that the server's certificate is trusted. If you run `./gradlew startStandalone` without it, you'll encounter the following error:
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md [hunk 11]
```diff
@@ -69,23 +70,69 @@ Download the [docker-compose.yml](https://github.com/pravega/pravega/tree/master
...
 
-Clients can then connect to the controller at `${HOST_IP}:9090`.
+```java
```
[the remark this answers is not in the document]
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda  accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 3 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Edited

File: documentation/src/docs/deployment/run-local.md [hunk 12]
```diff
@@ -9,31 +9,31 @@ You may obtain a copy of the License at
 -->
 # Running Pravega
 
-Running in local machine allows us to get started using Pravega very quickly. Most of the options uses the standalone mode which is suitable for most of the development and testing purposes.
+
+Running in local machine allows us to get started using Pravega very quickly. Running Pravega in Standalone mode is suitable for development and testing Pravega applications.
```
Comment (Aparnarr) [inline documentation/src/docs/deployment/run-local.md]: Edited

File: documentation/src/docs/deployment/run-local.md [hunk 13]
```diff
@@ -9,31 +9,31 @@ You may obtain a copy of the License at
...
 
 The prerequisites for running in local machine is described below.
 
 ## Standalone Mode
 
 ### From Source
 
-The Pravega source code needs to be checked out to download the dependencies:
+The Pravega source code needs to be checked out:
```
Comment (Aparnarr) [inline documentation/src/docs/deployment/run-local.md]: edited

File: documentation/src/docs/deployment/run-local.md [hunk 14]
```diff
@@ -76,9 +76,12 @@ HOST_IP=1.2.3.4 docker-compose up
 ```
 Clients can then connect to the controller at `${HOST_IP}:9090`.
 
-## Configuring standalone
+## Running Pravega in Standalone Mode with SSL/TLS Enabled
```
Comment (Aparnarr) [inline documentation/src/docs/deployment/run-local.md]: section title included

**Comment by Aparnarr** [comment]
@ravisharda All the requested changes have been implemented please check and record your further comments.

File: documentation/src/docs/deployment/run-local.md [hunk 15]
```diff
@@ -9,31 +9,31 @@ You may obtain a copy of the License at
...
 
 ```
 ./gradlew startStandalone
 ```
 
 ### From Installation Package
 
-Download the Pravega latest release from the [Github Releases](https://github.com/pravega/pravega/releases). The tarball or zip files can be used as they are identical. Instructions are provided for the tar files, but the same can be used for the zip file also.
-
+Download the Pravega latest release from the [GitHub Releases](https://github.com/pravega/pravega/releases). The tarball or zip files can be used as they are identical. Instructions are provided for the tar files, but the same can be used for the zip file also.
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: "Download the latest release of Pravega..." sounds better to me. 

About the sentences: "The tarball or zip files can be used as they are identical. Instructions are provided for the tar files, but the same can be used for the zip file also."  
* I'd suggest using the active voice like the previous sentence which starts with "Download the latest release..."
* "The tarball or zip files can be used as they are identical" looks incomplete to me. Perhaps the word `either` is missing.
* The sentences sound be made simpler: "Extract either the tarball or the zip file to a ...."

File: documentation/src/docs/deployment/run-local.md (hunk 13 above)
Comment (ravisharda) [reply documentation/src/docs/deployment/run-local.md]: One more thing here. Can we use active voice here (and everywhere else in this document): "Download the source code to your computer."?

File: documentation/src/docs/deployment/run-local.md [hunk 16]
```diff
@@ -76,9 +76,12 @@ HOST_IP=1.2.3.4 docker-compose up
 ```
 Clients can then connect to the controller at `${HOST_IP}:9090`.
 
-## Configuring standalone
+## Running Pravega in Standalone Mode with SSL/TLS Enabled
+
+By default both the `singlenode.enableTls` and `singlenode.enableauth` are disabled. The configurations, `singlenode.enableTls` and `singlenode.enableauth` can be used to enable encryption and authentication respectively.
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Since this subsection is now about TLS, references to auth related configuration should be avoided here, I believe. What do you say?

Comment (ravisharda) [reply documentation/src/docs/deployment/run-local.md]: Also, these sentences can be simplified: "By default, SSL/TLS is disabled. To enable it, set the value of  `singlenode.enableTls` to true. If you enable TLS...., you must also set the following configuration properties: ...."

File: documentation/src/docs/deployment/run-local.md [hunk 17]
```diff
@@ -132,7 +139,6 @@ Caused by: sun.security.validator.ValidatorException: PKIX path building failed:
                  .validateHostName(false)
                  .build();
     ```
-    Everything else should be the same as other reader/writer apps.
-
+6. Everything else should be the same as the Reader/Writer apps.
 
-6. Clients can then connect to the controller at `${HOST_IP}:9090`.
+These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Do we need this statement here? I don't think it is relevant for the user. What do you say?

File: documentation/src/docs/deployment/run-local.md (hunk 13 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: Accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 15 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: Accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 16 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 17 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: Accepted sugesstion

**Comment by Aparnarr** [comment]
@ravisharda  Please review and record if u have any further comments.

File: documentation/src/docs/deployment/run-local.md [hunk 18]
```diff
@@ -76,9 +79,13 @@ HOST_IP=1.2.3.4 docker-compose up
 ```
 Clients can then connect to the controller at `${HOST_IP}:9090`.
 
-## Configuring standalone
+## Running Pravega in Standalone Mode with SSL/TLS Enabled
 
-Configure standalone server to communicate using SSL/TLS. To do so, edit the TLS-related properties in `standalone-config.properties` as shown below:
+By default the `singlenode.enableTls` (SSl/TLS) is disabled. To enable it, set the value of  `singlenode.enableTls` to **True**. Once the `enableTls` (TLS) is enabled, the default certificates provided in the `conf` directory are used for setting up TLS.
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Can we rephrase this to something like: SSL/TLS is disabled by default in Pravega standalone mode cluster.

Also, item 1 in the points below already explains how to enable it, so the following statements are redundant: "To enable it, set the value of  `singlenode.enableTls` to **True**. Once the `enableTls` (TLS) is enabled, the default certificates provided in the `conf` directory are used for setting up TLS." We could say something like the following instead:

"The steps below show how to enable and run standalone mode cluster with SSL/TLS enabled:"

File: documentation/src/docs/deployment/run-local.md [hunk 19]
```diff
@@ -89,29 +96,33 @@ Configure standalone server to communicate using SSL/TLS. To do so, edit the TLS
   singlenode.trustStoreJKS=../config/standalone.truststore.jks
 
   ```
-These properties include ports for Zookeeper, Segment Store and Controller. They also contain other configurations related to security.
 
-## Running standalone with encryption and authentication enabled
+2. Ensure that the server's certificate is trusted. If you run `./gradlew startStandalone` without it, you'll encounter the following error:
+
+  ```java
+  Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Please remove reference to this exception, as suggested earlier.

File: documentation/src/docs/deployment/run-local.md [hunk 20]
```diff
@@ -76,9 +79,13 @@ HOST_IP=1.2.3.4 docker-compose up
...
 
-Configure standalone server to communicate using SSL/TLS. To do so, edit the TLS-related properties in `standalone-config.properties` as shown below:
+By default the `singlenode.enableTls` (SSl/TLS) is disabled. To enable it, set the value of  `singlenode.enableTls` to **True**. Once the `enableTls` (TLS) is enabled, the default certificates provided in the `conf` directory are used for setting up TLS.
+
+Set the following configuration properties:
+
+1. Configure standalone server to communicate using SSL/TLS. To do so, edit the TLS-related properties in `standalone-config.properties` as shown below:
 
   ```java
   singlenode.enableTls=true
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Please correct the indentation for the text and code under these numbered items of the list. For example, the formatted configuration code should start at the same location where the text for the numbered item starts.

Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Sure will edit

File: documentation/src/docs/deployment/run-local.md (hunk 18 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda Accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 20 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: Accepted suggestion

File: documentation/src/docs/deployment/run-local.md (hunk 19 above)
Comment (Aparnarr) [reply documentation/src/docs/deployment/run-local.md]: @ravisharda I am not clear with your comment.! What `references` are made?? not clear...u mean the blue font....if so (that is rendered for java format)!

**Comment by Aparnarr** [comment]
@ravisharda All the requested changes have been implemented please check and record your further comments.

File: documentation/src/docs/deployment/run-local.md (hunk 19 above)
Comment (ravisharda) [reply documentation/src/docs/deployment/run-local.md]: What I meant was we should remove the exception from the documentation. Please remove the following: 

```
If you run `./gradlew startStandalone` without it, you'll encounter the following error:

 Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
...
```

File: documentation/src/docs/deployment/run-local.md (hunk 20 above)
Comment (ravisharda) [reply documentation/src/docs/deployment/run-local.md]: Indentation is not corrected in the latest commit.

**Comment by Aparnarr** [comment]
@ravisharda Addressed your comments. Please check

File: documentation/src/docs/deployment/run-local.md [hunk 21]
```diff
@@ -71,68 +72,59 @@ wget https://raw.githubusercontent.com/pravega/pravega/master/docker/compose/doc
...
-  `sudo keytool -importcert -alias local-CA -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts  -file cert.der` (using the default password 'changeit')
+   **Note:** If you want to use a custom truststore instead of adding the certificate to the system truststore, create a new truststore using Java keytool utility, add the certificate to it and configure the JVM to use it by setting the system properties `javax.net.ssl.trustStore` and `javax.net.ssl.trustStorePassword`.
 
 3. Run Pravega standalone mode using command `./gradlew startStandalone`.
 
 4. Verify that controller REST API is returning response over SSL/TLS:
 
     ```java
-     curl -v -k https://104.215.152.115:9091/v1/scopes
+      curl -v -k https://104.215.152.115:9091/v1/scopes
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: Please generalize this URL to something like (`https://<host-ip>:....)) to avoid putting the specific IP address.

File: documentation/src/docs/deployment/run-local.md [hunk 22]
```diff
@@ -71,68 +72,59 @@ wget https://raw.githubusercontent.com/pravega/pravega/master/docker/compose/doc
...
-        at sun.security.validator.Validator.validate(Validator.java:260)
-        at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManager
-```
-2. To ensure the server's certificate is trusted, import it into the JVM's truststore. The following command sequence is used (in Linux) with the provided certificate file "cert.pem":
+The following command sequence is used (in Linux) with the provided certificate file `cert.pem` into the JVM's system truststore.
+
+   - `cd /path/to/pravega/config`
+   - Convert the `cert.pem` file to `DER` format: `openssl x509 -in cert.pem -inform pem -out cert.der  -outform der`
+   - Import the certificate into the local JVM's trust store:
+    `sudo keytool -importcert -alias local-CA -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts  -file cert.der` (using the default password `changeit`)
```
Comment (ravisharda) [inline documentation/src/docs/deployment/run-local.md]: please generalize the path to JRE here, as it can vary from one machine to another. Something like `/path/to/jre/lib/security/cacerts` works better here.

**Comment by ravisharda** [comment]
LGTM.

**Comment by Aparnarr** [comment]
@fpj @andreipaduroiu Please review and proceed for merging

**Comment by Aparnarr** [comment]
All comments addressed

**Comment by Aparnarr** [comment]
@andreipaduroiu this file is done....please proceed with merging.
